### PR TITLE
Make ClusterNetworkingConfig resource dynamically configurable

### PR DESCRIPTION
**Pull request by andrejtokarcik**
This PR turns `types.ClusterNetworkingConfig` (introduced in #6638) into a [dynamically configurable resource](https://github.com/gravitational/teleport/blob/master/rfd/0016-dynamic-configuration.md).

### CODE REVIEW

File: lib/auth/grpcserver_test.go [hunk 1]
```diff
@@ -1018,3 +1018,59 @@ func TestDeleteLastMFADevice(t *testing.T) {
 		checkErr: require.Error,
 	})
 }
+
+// TestClusterNetworkingConfigOriginDynamic tests setting ClusterNetworkingConfig
+// via gRPC results in configuration with OriginDynamic being stored.
+func TestClusterNetworkingConfigOriginDynamic(t *testing.T) {
+	ctx := context.Background()
```
Comment (Joerger) [inline lib/auth/grpcserver_test.go:1025]:
```suggestion
        t.Parallel()
	ctx := context.Background()
```
Can this run parallel to other tests?

File: lib/auth/init_test.go [hunk 2]
```diff
@@ -154,158 +154,211 @@ func TestBadIdentity(t *testing.T) {
...
 }
 
-func TestAuthPreferenceInitFromConfigFileToDefault(t *testing.T) {
-	// Simulate a server with auth preference from config file.
-	var err error
-	conf := setupConfig(t)
-	conf.AuthPreference, err = types.NewAuthPreferenceFromConfigFile(types.AuthPreferenceSpecV2{
-		SecondFactor: constants.SecondFactorOff,
+func testDynamicallyConfigurable(t *testing.T, p testDynamicallyConfigurableParams) {
+	initAuthServer := func(conf InitConfig) *Server {
```
Comment (Joerger) [inline lib/auth/init_test.go:164]: Should `initAuthServer` take in the `t *testing.T` parameter?

Comment (andrejtokarcik) [reply lib/auth/init_test.go:164]: Well, I don't know, should it? :) The only relevant `t` is captured by the closure so the code apparently works as is.

Comment (Joerger) [reply lib/auth/init_test.go:164]: Sorry that was a vague comment as I wasn't positive about the following. `t.Run` spawns a child `testing.T`, so the `t.Cleanup` in `initAuthServer` doesn't run until `testDynamicallyConfigurable` returns. If you pass the `testing.T` spawned by `t.Run`, the cleanup will occur immediately after `t.Run` concludes. It's a small performance improvement that will matter more in other tests.

File: tool/tctl/common/resource_command.go [hunk 3]
```diff
@@ -444,9 +445,43 @@ func (rc *ResourceCommand) createAuthPreference(client auth.ClientI, raw service
...
+
+	exists := storedNetConfig.Origin() != types.OriginDefaults
+	if !rc.force && exists {
+		return trace.AlreadyExists("non-default cluster networking configuration already exists")
+	}
+
+	managedByStaticConfig := storedNetConfig.Origin() == types.OriginConfigFile
+	if !rc.confirm && managedByStaticConfig {
+		return trace.BadParameter(managedByStaticCreateMsg)
+	}
```
Comment (awly) [inline tool/tctl/common/resource_command.go:470]: nit: extract this into a helper and reuse in `createAuthPreference`

File: tool/tctl/common/resource_command.go [hunk 4]
```diff
@@ -817,14 +868,24 @@ func (rc *ResourceCommand) getCollection(client auth.ClientI) (ResourceCollectio
 		return nil, trace.NotFound("kube_service with ID %q not found", rc.ref.Name)
 	case services.KindClusterAuthPreference:
 		if rc.ref.Name != "" {
-			return nil, trace.BadParameter("only simple `tctl get cluster_auth_preference` can be used")
+			return nil, trace.BadParameter("only simple `tctl get %v` can be used", services.KindClusterAuthPreference)
```
Comment (awly) [inline tool/tctl/common/resource_command.go:871]:
```suggestion
			return nil, trace.BadParameter("only simple `tctl get %v` can be used", types.KindClusterAuthPreference)
```

File: tool/tctl/common/collection.go [hunk 5]
```diff
@@ -499,6 +499,31 @@ func (c *authPrefCollection) writeText(w io.Writer) error {
 	return trace.Wrap(err)
 }
 
+type netConfigCollection struct {
+	netConfigs []types.ClusterNetworkingConfig
```
Comment (awly) [inline tool/tctl/common/collection.go:503]: there can only be one `ClusterNetworkingConfig`
don't make this a slice

File: lib/auth/init_test.go (hunk 2 above)
Comment (awly) [reply lib/auth/init_test.go:164]: please pass `testing.T` as an argument here
the closure works, but any failure will be associated with the top-level test, not the subtest
passing a subtest-specific `t` will make the failure clearer

**Review by awly — APPROVED** [review]

**Review by Joerger — APPROVED** [review]

**Review by russjones — APPROVED** [review]
Bot.

**Pull request merged**